### PR TITLE
[14012, 14171] Implement RTPS extensions for writer-side filtering

### PR DESCRIPTION
**Pull request by MiguelCompany**
## Description
Just what the title says.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [x] **N/A** Documentation builds and tests pass locally. 
- [x] **N/A** New feature has been added to the `versions.md` file (if applicable).
- [x] **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by JLBuenoLopez-eProsima** [comment]
Linters have failed so I have unchecked the code style checkbox. Please, recheck after fixing.

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]
CI failures seem unrelated.

File: include/fastdds/dds/core/policy/ParameterTypes.hpp [hunk 1]
```diff
@@ -56,76 +56,103 @@ namespace dds {
...
-    PID_PARTICIPANT_SECURITY_INFO = 0x1005,
-    PID_IDENTITY_STATUS_TOKEN = 0x1006,
-    PID_PERSISTENCE_GUID = 0x8002,
-    PID_RELATED_SAMPLE_IDENTITY = 0x800f,
-    PID_DATA_REPRESENTATION = 0x0073,
-    PID_TYPE_CONSISTENCY_ENFORCEMENT = 0x0074,
-    PID_TYPE_INFORMATION = 0x0075,
-    PID_DISABLE_POSITIVE_ACKS = 0x8005,
-    PID_DATASHARING = 0x8006,
+    /* From Table 9.18 of DDS-RTPS 2.5 */
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/dds/core/policy/ParameterTypes.hpp:59]: The uncrustify job seems to be happy with the current changes, so I guess I can ignore this comment?

[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply include/fastdds/dds/core/policy/ParameterTypes.hpp:59]: Yes, I think the indentation was the reason but it seems there was something else.

**Comment by MiguelCompany** [comment]
@JLBuenoLopez-eProsima I applied the suggestions on the src files. The tests are still pending.

File: test/blackbox/common/RTPSBlackboxTestsDiscovery.cpp [hunk 2]
```diff
@@ -0,0 +1,950 @@
...
+    // Test second iteration: expect ReaderDiscoveryInfo::CHANGED_QOS_READER.
+    reader.update();
+    {
+        std::unique_lock<std::mutex> lock(mutex);
+        cv.wait(lock, [&iteration]()
+                {
+                    return Iterations::CHANGED_QOS_READER == iteration || Iterations::WITH_ERROR == iteration;
+                });
+        ASSERT_EQ(Iterations::CHANGED_QOS_READER, iteration);
+        ASSERT_EQ(0, content_filter_property.content_filtered_topic_name.size());
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/blackbox/common/RTPSBlackboxTestsDiscovery.cpp:828]: I think `0` is a special case. I have never seen it on warnings. Specially when compared with a `size_t`.

File: test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp [hunk 3]
```diff
@@ -288,6 +337,813 @@ TEST(BuiltinDataSerializationTests, ignore_unsupported_type_object)
...
+        EXPECT_TRUE(fastdds::dds::ParameterSerializer<Parameter_t>::add_parameter_sentinel(&msg));
+
+        msg.pos = 0;
+        EXPECT_TRUE(out.readFromCDRMessage(&msg, network, true));
+
+        ASSERT_EQ("", out.content_filter().content_filtered_topic_name.to_string());
+        ASSERT_EQ("", out.content_filter().related_topic_name.to_string());
+        ASSERT_EQ("", out.content_filter().filter_class_name.to_string());
+        ASSERT_EQ("", out.content_filter().filter_expression);
+        ASSERT_EQ(0, out.content_filter().expression_parameters.size());
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp]: See 291f426

File: test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp [hunk 4]
```diff
@@ -288,6 +337,813 @@ TEST(BuiltinDataSerializationTests, ignore_unsupported_type_object)
...
+    }
+}
+
+/*!
+ * \test RTPS-CFT-CFP-07 Tests serialization of `ContentFilterProperty_t` when `filterClassName` is empty.
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp:701]: Added to the doxydoc of the fields on d5ddcd2

File: test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp [hunk 5]
```diff
@@ -288,6 +337,813 @@ TEST(BuiltinDataSerializationTests, ignore_unsupported_type_object)
...
+    ASSERT_EQ(in.content_filter().filter_expression, out.content_filter().filter_expression);
+    ASSERT_EQ(in.content_filter().expression_parameters.size(), out.content_filter().expression_parameters.size());
+    ASSERT_EQ(in.content_filter().expression_parameters[0], out.content_filter().expression_parameters[0]);
+    ASSERT_EQ(in.content_filter().expression_parameters[1], out.content_filter().expression_parameters[1]);
+    ASSERT_EQ(in.content_filter().expression_parameters[2], out.content_filter().expression_parameters[2]);
+    ASSERT_EQ(in.content_filter().expression_parameters[3], out.content_filter().expression_parameters[3]);
+}
+
+/*!
+ * \test RTPS-CFT-CFP-03 Tests serialization of `ContentFilterProperty_t` fails with a wrong `contentFilteredTopicName`.
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/unittest/rtps/builtin/BuiltinDataSerializationTests.cpp:433]: The test plan is using the names on the RTPS standard. I think we should stick to them in the comments. Same elsewhere.

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM :+1:

**Comment by JLBuenoLopez-eProsima** [comment]
CI failures seem unrelated. The tests that need to be introduced I think they are being done in #2621. Consequently, I think this PR is ready to be merged in the intermediary branch.

**Pull request merged**